spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpOutboundGateway.java

```diff
@@ -155,6 +156,9 @@ protected Object handleRequestMessage(Message<?> requestMessage) {
 					logger.debug("released semaphore");
 				}
 			}
+			if (this.connectionFactory instanceof CloseDeferrable) {
+				((CloseDeferrable) this.connectionFactory).closeDeferred(connectionId);
```

Comment:
OK. I worried how we closed a connection before (or even right now artificially) and found that `CachedConnection` does that itself in the `onMessage`.

So, merging as is, but we reall should rethink how it works and who is responsible for what.